### PR TITLE
Only take the upper bounds of modes associated with allocations

### PR DESCRIPTION
**Pull request by lpw25**
Rather than taking the upper bound of modes when translating them in translcore, we instead only take the upper bound of modes associated with allocations, which we do between typing and translation to lambda. For all other modes we take the lower bound. This will allow us to generate fewer regions.

This PR is on top of #35 and by extension #34

### CODE REVIEW

File: typing/typecore.ml [hunk 1]
```diff
@@ -2913,6 +2935,8 @@ and type_expect_
     end
   | Pexp_constant cst ->
       let cst = constant_or_raise env loc cst in
+      if is_allocated_constant cst then
+        register_allocation expected_mode;
```
Comment (stedolan) [inline typing/typecore.ml]: Is there any particular reason to want constants to be in mode Local?

Comment (lpw25) [reply typing/typecore.ml]: Probably not, but I'd rather just treat all allocations uniformly than reason about which ones are guaranteed to be statically allocated, at least whilst the cost is insignificant.

(I suppose that if you use `-unsafe-string` then it still matters for strings, but I don't think anyone cares about that)

Comment (stedolan) [reply typing/typecore.ml]: `-unsafe-string` isn't even an option any more, FWIW.

**Review by stedolan — CHANGES_REQUESTED** [review]
Looks good, but the Pexp_constant / is_allocated_constant stuff should be removed. Texp_constant is not an allocation.

**Pull request merged**